ycmd-eldoc.el

```diff
@@ -0,0 +1,119 @@
+;;; ycmd-eldoc.el --- Eldoc support for ycmd-mode    -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2016  Peter Vasil
+
+;; Author: Peter Vasil <mail@petervasil.net>
+;; URL: https://github.com/abingham/emacs-ycmd
+;; Version: 0.1
+;; Package-Requires: ((ycmd "0.1") (deferred "0.2.0") (s "1.9.0") (dash "1.2.0") (cl-lib "0.5"))
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; To use this package, add these lines to your init.el file:
+;;
+;;     (require 'ycmd-eldoc)
+;;     (add-hook 'ycmd-mode-hook 'ycmd-eldoc-setup)
+;;
+
+;;; Code:
+
+(require 'cl-lib)
+(require 'eldoc)
+(require 'ycmd)
+(require 'deferred)
+(require 'dash)
+(require 's)
+
+
+(defvar-local ycmd-eldoc--cache nil)
+
+(defun ycmd-eldoc--documentation-function ()
+  "Eldoc function for `ycmd-mode'."
+  (when ycmd-mode
+    (--when-let (ycmd-eldoc--info-at-point)
+      (eldoc-message it))))
+
+(defun ycmd-eldoc--info-at-point ()
+  "Get function info at point."
+  (save-excursion
+    (ycmd-eldoc--goto-func-name)
+    (-when-let (symbol (symbol-at-point))
+      (if (eq symbol (car ycmd-eldoc--cache))
+          (cadr ycmd-eldoc--cache)
+        (deferred:sync!
```

Comment:
Need to use the deferred call with sync because `eldoc` is calling the `eldoc-documentation-function` already async with a timer. If we don't sync here the output message is the content of the deferred object.

@abingham Would be great if you could take a look. Maybe we can sort this out.